### PR TITLE
various openrct2.d.ts improvements

### PR DESCRIPTION
**Pull request by Sadret**
Adds the exact type to each TileElement, overwriting the more general type of BaseTileElement.

Edit:
- add TileElement union type
- add the exact type to each TileElement, overwriting the more general type of BaseTileElement
- use TileElement type in Tile interface (instead of BaseTileElement)
- change single quotes to double quotes
- add type to WidgetBase

### CODE REVIEW

**Review by IntelOrca — APPROVED** [review]

**Review by IntelOrca — CHANGES_REQUESTED** [review]
I think `type` needs to be removed from `BaseTileElement`. I remember having to do that for the Widget types.

**Comment by Sadret** [comment]
Why do you think so? My local IDE does not report any problems.
Also, any BaseTileElement should have a type, so why remove it?

**Comment by IntelOrca** [comment]
I am not quite sure. I just remember needing to do it for Widget, otherwise I couldn't define an array of widgets where the type field was enough to say what widget type it was.

**Comment by Sadret** [comment]
But we don't have a discriminated union type for TileElements, so that cannot be a problem.

**Comment by Sadret** [comment]
Especially, I think we should reintroduce `WidgetBase.type`. This code compiles fine with the `type` field included:
```
function typeTest(widget: Widget) {
    if (widget.type === "button")
        console.log(widget.border);
}
```

**Comment by Sadret** [comment]
Maybe you tested against `WidgetBase` instead of `Widget`?

**Comment by IntelOrca** [comment]
No I think was when I had an array of widgets defined:
```
widgets: [
    { type: 'button', etc. },
    { type: 'label', etc. },
]
```

**Comment by Sadret** [comment]
Maybe the type inference just failed and assumed `WidgetBase` instead of `Widget`?

**Comment by Sadret** [comment]
I have two other things that I think should be changed:

- add a `TileElement` union type and use it where it fits (similiar to `Widget` vs `WidgetBase`), e.g. in the `Tile` interface.
```
    type TileElement =
        SurfaceElement | FootpathElement | TrackElement | SmallSceneryElement | WallElement | EntranceElement
        | LargeSceneryElement | BannerElement | CorruptElement;
```

- use " (double quotes) consistently instead of ' (single quotes) (as double quotes is the majority right now)

If that's ok, I will modify the PR to include this and `WidgetBase.type`.

**Comment by IntelOrca** [comment]
> If that's ok, I will modify the PR to include this and WidgetBase.type.

I will need to find out why I couldn't do this. In the meantime I am happy for you to do the other changes.

File: distribution/changelog.txt [hunk 1]
```diff
@@ -59,6 +59,7 @@
 - Improved: [#6022] Allow up to 128 ride objects to be selected in track designer.
 - Improved: [#12917] Changed peep movement so that they stay more spread out over the full width of single tile paths.
 - Improved: [#13386] A GUI error message is now displayed if the language files are missing.
+- Improved: [#14193] Various openrct2.d.ts improvements.
```
Comment (tupaschoal) [inline distribution/changelog.txt:62]: I'd rather have multiple lines with the same PR number than having a vague changelog entry.

**Review by IntelOrca — APPROVED** [review]

File: distribution/changelog.txt (hunk 1 above)
Comment (Sadret) [reply distribution/changelog.txt]: I grouped the changes into three entries.

**Pull request merged**